src/Core/Form/IdentifiableObject/DataHandler/AbstractCategoryFormDataHandler.php

```diff
@@ -0,0 +1,183 @@
+        $menuThumbnailImages = $data['menu_thumbnail_images'] ?? [];
+        $this->uploadImages(
+            $categoryId,
+            $data['cover_image'],
+            $data['thumbnail_image'],
+            $menuThumbnailImages
+        );
+
+        return $categoryId->getValue();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update($categoryId, array $data)
+    {
+        $availableKeys = $this->getAvailableKeys((int) $categoryId);
+
+        if (!isset($data['menu_thumbnail_images']) || count($data['menu_thumbnail_images']) > count($availableKeys)) {
+            throw new MenuThumbnailsLimitException(sprintf('Maximum number of menu thumbnails was reached for category "%d"', $categoryId));
+        }
+        $command = $this->createEditCategoryCommand($categoryId, $data);
+
+        $this->commandBus->handle($command);
+        $categoryId = new CategoryId((int) $categoryId);
+
+        /**
+         * In some cases in form menu_thumbnail_images can be disabled so value won't get here.
+         */
+        $menuThumbnailImages = $data['menu_thumbnail_images'] ?? [];
+
+        $this->uploadImages(
+            $categoryId,
+            $data['cover_image'],
+            $data['thumbnail_image'],
+            $menuThumbnailImages
+        );
+    }
+
+    /**
+     * @param CategoryId $categoryId
+     * @param UploadedFile|null $coverImage
+     * @param UploadedFile|null $thumbnailImage
+     * @param UploadedFile[] $menuThumbnailImages
+     */
+    protected function uploadImages(
+        CategoryId $categoryId,
+        UploadedFile $coverImage = null,
+        UploadedFile $thumbnailImage = null,
+        array $menuThumbnailImages = []
+    ) {
+        if (null !== $coverImage) {
+            $this->categoryCoverUploader->upload($categoryId->getValue(), $coverImage);
+        }
+
+        if (null !== $thumbnailImage) {
+            $this->categoryThumbnailUploader->upload($categoryId->getValue(), $thumbnailImage);
+        }
+
+        if (!empty($menuThumbnailImages)) {
+            foreach ($menuThumbnailImages as $menuThumbnail) {
+                $this->categoryMenuThumbnailUploader->upload($categoryId->getValue(), $menuThumbnail);
+            }
+        }
+    }
+
+    /**
+     * @param int $categoryId
+     *
+     * @return array<int, int>
+     */
+    protected function getAvailableKeys(int $categoryId): array
+    {
+        $files = scandir(_PS_CAT_IMG_DIR_, SCANDIR_SORT_NONE);
+        $usedKeys = [];
+
+        foreach ($files as $file) {
+            $matches = [];
+
+            if (preg_match('/^' . $categoryId . '-([0-9])?_thumb.jpg/i', $file, $matches) === 1) {
+                $usedKeys[] = (int) $matches[1];
+            }
+        }
+
+        return array_diff(MenuThumbnailId::ALLOWED_ID_VALUES, $usedKeys);
+    }
+
+    abstract protected function createAddCategoryCommand(array $data);
+
+    abstract protected function createEditCategoryCommand($categoryId, array $data);
```

Comment:
Return values is missing
Type for $categoryId is missing

Reply:
Abstract add/addroot and edit/editroot category commands so return value can be added.